### PR TITLE
 [THREESCALE-2194] Ensure locks are released on configuration_loader. 

### LINKED ISSUE #1010: Ensure lock is released in the ConfigurationLoader module
When downloading the configuration from Porta, we use an `ngx.semaphore` to make sure that there are not several requests at the same time trying to download the same configuration and storing it in memory.

Here's the code function that does that:
https://github.com/3scale/APIcast/blob/6f6aeb034e0c388c9280c6e259f4ff9303be859f/gateway/src/apicast/configuration_loader.lua#L199

The problem is that if there's an error while trying to download the config, the thread dies and the lock is never returned. Subsequent requests that try to download the config for the same service will wait for 15 s and then fail because they will not be able to acquire the lock (it was never returned in the first call). There's a bug reported when this happened:  https://issues.jboss.org/browse/THREESCALE-2194

My first thought was to just wrap that code on an xpcall or pcall. However, I think that we should implement a generic "finally/ensure" clause because we have the same pattern in different parts of the code. For example: https://github.com/3scale/APIcast/blob/e9c8f0a59318fa4f37e9fa41422cce5d046a92ff/gateway/src/apicast/policy/3scale_batcher/3scale_batcher.lua#L112

### PR DESCRIPTION
**Pull request by eloycoto**
This commits add a new function `run` to the synchronization module,
that function ensures that a given function is going to be run in a safe
task and make sure that the release of the lock  always happens, so the
lock is not going to be blocked if something goes wrong.

Fix #1010
Fix THREESCALE-2194

### CODE REVIEW

File: t/configuration-loading-lazy.t [hunk 1]
```diff
@@ -75,7 +76,7 @@ env THREESCALE_PORTAL_ENDPOINT=http://127.0.0.1:$TEST_NGINX_SERVER_PORT;
 GET /t?user_key=fake
 --- error_code: 200
 --- user_files eval
-[
+[
```
Comment (davidor) [inline t/configuration-loading-lazy.t]: :eyes:

File: gateway/src/resty/synchronization.lua [hunk 2]
```diff
@@ -53,4 +53,28 @@ function _M:release(key)
   semaphores[key] = nil
 end
 
+--- run a new function, callback using locks on the given key
+-- @tparam string key: key for the semaphore
+-- @tparam int timeout: timeout for getting the lock before raise an error.
+-- @tparam function callback: function to execute if the lock is adquired correctly
+function _M:run(key, timeout, callback, ...)
+  local sema = self:acquire(key)
```
Comment (davidor) [inline gateway/src/resty/synchronization.lua]: `acquire` can return an error.

File: gateway/src/resty/synchronization.lua [hunk 3]
```diff
@@ -53,4 +53,28 @@ function _M:release(key)
   semaphores[key] = nil
 end
 
+--- run a new function, callback using locks on the given key
+-- @tparam string key: key for the semaphore
+-- @tparam int timeout: timeout for getting the lock before raise an error.
+-- @tparam function callback: function to execute if the lock is adquired correctly
```
Comment (davidor) [inline gateway/src/resty/synchronization.lua]: typo acquired

Comment (davidor) [inline gateway/src/resty/synchronization.lua]: Maybe we should also add `...` to the documentation of the function. I can't remember the syntax now.

File: gateway/src/resty/synchronization.lua [hunk 4]
```diff
@@ -53,4 +53,28 @@ function _M:release(key)
...
+  local ok, err = sema:wait(timeout)
+  if not ok then
+    ngx.log(ngx.WARN, 'failed to acquire lock on key: ', key, ' error: ', err)
+    return false, nil
+  end
+
+  local task = safe_task.new(callback)
+  local ret, result = task:execute(...)
+
+  if ok then
```
Comment (davidor) [inline gateway/src/resty/synchronization.lua]: Can we name this `lock_acquired` or similar? I think that would make it perfectly clear what this refers to.

File: gateway/src/resty/synchronization.lua [hunk 5]
```diff
@@ -53,4 +53,28 @@ function _M:release(key)
...
+    self.release(key)
+    sema:post()
+  end
+
+  return ret, result
```
Comment (davidor) [inline gateway/src/resty/synchronization.lua]: Are 2 enough? According to the docs of the function, `execute` returns 2 or 3 params depending on whether there was an error or not.

File: t/configuration-loading-lazy.t [hunk 6]
```diff
@@ -95,3 +96,30 @@ GET /t?user_key=fake
   | ]
 ]
 
+=== TEST 4: load invalid json
+To validate that process does not died with invalid config
```
Comment (davidor) [inline t/configuration-loading-lazy.t]: I'm not sure this test is what we need. I checked that it passes without applying the patch. Also, this description seems to contradict the one in TEST 2.

File: spec/synchronization.lua [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+    synchronization = require('resty.synchronization').new(1)
+  end)
+
+  it("Valid run execution", function()
+    local callback = spy.new(function() return 1 end)
+
+    local ret, result = synchronization:run(key, 10, callback)
+    assert.same(ret, true)
+    assert.same(result, {1})
+    assert.spy(callback).was.called()
```
Comment (davidor) [inline spec/synchronization.lua]: o we need the spies? I think that checking the results of `run` might be enough.

File: gateway/src/apicast/configuration_loader.lua [hunk 8]
```diff
@@ -196,33 +196,30 @@ function lazy.init(configuration)
...
+  if _M.configured(configuration, host) then
+    return configuration
   end
 
-  if ok then
-    synchronization:release(host)
-    sema:post()
-  else
-    ngx.log(ngx.WARN, 'failed to acquire lock to lazy load: ', host, ' error: ', err)
+  local ret, result = synchronization:run(host, 15, lazy.load_config, configuration, host)
```
Comment (davidor) [inline gateway/src/apicast/configuration_loader.lua]: Can we define a constant with this 15? I know it's not a problem introduced in this PR, but looks like a good opportunity to fix this.

File: gateway/src/apicast/configuration_loader.lua [hunk 9]
```diff
@@ -196,33 +196,30 @@ function lazy.init(configuration)
   configuration.configured = true
 end
 
+function lazy.load_config(configuration, host)
```
Comment (davidor) [inline gateway/src/apicast/configuration_loader.lua]: This method is exposed. It should be private, right?

File: gateway/src/apicast/configuration_loader.lua [hunk 10]
```diff
@@ -196,33 +196,30 @@ function lazy.init(configuration)
...
   end
 
-  if ok then
-    synchronization:release(host)
-    sema:post()
-  else
-    ngx.log(ngx.WARN, 'failed to acquire lock to lazy load: ', host, ' error: ', err)
+  local ret, result = synchronization:run(host, 15, lazy.load_config, configuration, host)
+  if not ret then
+    ngx.log(ngx.WARN, 'failed to acquire lock to lazy load: ', host, ' error: ', result)
```
Comment (davidor) [inline gateway/src/apicast/configuration_loader.lua]: this error message is also shown in the synchronization module. Maybe here we should show a more specific error here.

File: spec/synchronization.lua [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+  it("Validates that run send correctly the arguments", function()
+    local callback = spy.new(function(key, value) return key, value end)
+
+    local ret, result = synchronization:run(key, 4, callback, "bob", "alice")
+    assert.same(ret, true)
+    assert.same(result, {"bob", "alice"})
+    assert.spy(callback).was_called()
+  end)
+
+  it("Validates that run does not breaks if a execption on the callback", function()
```
Comment (davidor) [inline spec/synchronization.lua]: typos: `break` `exception`.

File: spec/synchronization.lua (hunk 7 above)
Comment (eloycoto) [reply spec/synchronization.lua]: It's just to make the test completed and easy to understand if someone review the code in future.

File: t/configuration-loading-lazy.t (hunk 1 above)
Comment (eloycoto) [reply t/configuration-loading-lazy.t]: just empty whitespace on the end of the line. I think that it's good to have it clean :-)

File: t/configuration-loading-lazy.t (hunk 6 above)
Comment (eloycoto) [reply t/configuration-loading-lazy.t]: test 2 validates that the test returns a 404, but the json.decode of an empty string is not going to fail. But json decode of an invalid json will break.

Comment (eloycoto) [reply t/configuration-loading-lazy.t]: Made this test, looks more legitimate and it's fails: 
```
=== TEST 5: Invalid json structure
To validate that process does not died if the server return an invalid json configuration
--- main_config
env THREESCALE_PORTAL_ENDPOINT=http://127.0.0.1:$TEST_NGINX_SERVER_PORT/;
env APICAST_CONFIGURATION_LOADER=lazy;
--- http_config
  include $TEST_NGINX_UPSTREAM_CONFIG;
  lua_package_path "$TEST_NGINX_LUA_PATH";
--- config
  include $TEST_NGINX_APICAST_CONFIG;
  include $TEST_NGINX_BACKEND_CONFIG;

  location = /admin/api/nginx/spec.json {
    try_files /config.json =404;
  }

  location /api/ {
    echo "all ok";
  }
--- request
GET /t?user_key=fake
--- error_code: 404
--- user_files
>>> config.json
{
    "services": [{
      "id": 1,
      "backend_version": 1,
      "proxy": {
        "api_backend": ["test1", "test2"],
        "proxy_rules": [
          { "pattern": "/t", "http_method": "GET", "metric_system_name": "test" }
        ]
      }
    }]
}
```

File: spec/synchronization.lua (hunk 7 above)
Comment (davidor) [reply spec/synchronization.lua]: :+1:

File: gateway/src/apicast/configuration_loader.lua (hunk 9 above)
Comment (davidor) [reply gateway/src/apicast/configuration_loader.lua]: Just checking. Did you forget to change this or decided not to?

File: gateway/src/apicast/configuration_loader.lua [hunk 12]
```diff
@@ -9,7 +9,7 @@ local oidc_loader = require 'apicast.configuration_loader.oidc'
 local util = require 'apicast.util'
 local env = require('resty.env')
 local resty_url = require('resty.url')
-local synchronization = require('resty.synchronization').new(1)
+local synchronization = require('resty.synchronization').new(nil,1)
```
Comment (davidor) [inline gateway/src/apicast/configuration_loader.lua]: Why did this change?

File: gateway/src/resty/synchronization.lua [hunk 13]
```diff
@@ -53,4 +53,33 @@ function _M:release(key)
...
+--- run a new function, callback using locks on the given key
+-- @tparam string key: key for the semaphore
+-- @tparam int timeout: timeout for getting the lock before raise an error.
+-- @tparam function callback: function to execute if the lock is acquired correctly
+-- @param ...: the variable number of arguments that are going to be send to the callback function.
+function _M:run(key, timeout, callback, ...)
+  local sema, err = self:acquire(key)
+  if err ~= key then
+    ngx.log(ngx.WARN, 'failed to acquire lock on key: ', key, ' error: ', err)
+    return false, nil
```
Comment (davidor) [inline gateway/src/resty/synchronization.lua]: There's no need for the `, nil`

File: gateway/src/resty/synchronization.lua (hunk 5 above)
Comment (davidor) [reply gateway/src/resty/synchronization.lua]: When there's an error, `execute` returns 3 params and the error msg is the third one, so I think we'll lose it.
https://github.com/3scale/APIcast/blob/6b42baa85c9bde860c4c3bf70e8af78b48a710cf/gateway/src/resty/concurrent/safe_task_executor.lua#L39

File: spec/synchronization.lua [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+  it("Validates that run send correctly the arguments", function()
+    local callback = spy.new(function(key, value) return key, value end)
+
+    local ret, result = synchronization:run(key, 4, callback, "bob", "alice")
+    assert.same(ret, true)
+    assert.same(result, {"bob", "alice"})
+    assert.spy(callback).was_called()
+  end)
+
+  it("Validates that run does not break if a exception on the callback", function()
```
Comment (davidor) [inline spec/synchronization.lua]: but there's no exception in this test, right?

File: spec/synchronization.lua [hunk 15]
```diff
@@ -0,0 +1,78 @@
+describe("Synchronization", function()
```
Comment (davidor) [inline spec/synchronization.lua]: this file needs to be named `sychronization_spec.lua`. Without the `_spec` our Makefile will not run the test in this file.

**Pull request merged**